File: src/potemkin/LazyMapEntry.java
package potemkin;

import clojure.lang.ILookup;
import clojure.lang.MapEntry;

public class LazyMapEntry extends clojure.lang.MapEntry {
    private final Object _key;
    private final ILookup _map;

    public LazyMapEntry(ILookup map, Object key) {
        super(key, null);
        _key = key;
        _map = map;
    }

    @Override
    public Object val() {
        return _map.valAt(_key);
    }
}

File: src/potemkin/LazyMapEntry.clj
(ns potemkin.LazyMapEntry
  (:import (clojure.lang ILookup MapEntry))
  (:gen-class :extends clojure.lang.MapEntry
              :init init
              :constructors {[clojure.lang.ILookup Object] [Object Object]}
              :state state))

(defn -init [map k]
  [[k nil] map])

(defn -val [^potemkin.LazyMapEntry this]
  (.valAt ^ILookup (.state this) (.key this)))
